sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3163,8 +3163,11 @@ pi_result piDeviceGetInfo(pi_device Device, pi_device_info ParamName,
     for (auto &ZesMemHandle : MemHandles) {
       ZesStruct<zes_mem_properties_t> ZeMemProperties;
       ZE_CALL(zesMemoryGetProperties, (ZesMemHandle, &ZeMemProperties));
-      // Only report HBM which zeMemAllocDevice allocates from.
-      if (ZeMemProperties.type != ZES_MEM_TYPE_HBM)
+      // Only report device memory which zeMemAllocDevice allocates from.
```

Comment:
Ultimately, this query should return the free memory from where `zeMemAllocDevice` would allocate from. The previous guidance was to check for HBM, which is perhaps not always accurate. @jandres742 : how do we tell which memory the device allocations are made from?

Reply:
there is no HBM on Arc Alchemist devices hence I proposed to check if memory is -not- L1, L3, GRF or SLM since they're the types defined in L0 that can't be used for explicit allocations (https://github.com/oneapi-src/level-zero/blob/474188ae004a5c76953a829477997bc341e70d48/include/zes_api.h#L3174). 
It would have to be changed in case a new level of cache/EU memory/registers gets introduced in L0. 
I have looked for an API that would report the "main" type of memory and found none so this looks like the current best option. 
Having Arc Alchemist returning 0 while telling extension is supported as it's currently the case is more problematic since it may lead applications to avoid allocating anything on the device.

Reply:
I agree we need to fix this, just find it error-prone that this code is trying to "guess" the behavior of `zeMemAllocDevice`. We need a clean interface to know where exactly device allocations are made from.

Reply:
solution proposed by @ph0b  is cleaner, agree.  

> how do we tell which memory device allocations are made from

that is actually selected by users. In the zeMemAllocDevice descriptor you have an ordinal:

https://spec.oneapi.io/level-zero/latest/core/api.html#_CPPv4N26ze_device_mem_alloc_desc_t7ordinalE

The ordinal is supposed to be used for this, so if you have in your system HBM and DDR5 on the same devicer, then you would use the ordinal to select one or the other (value for ordinal depends on values returned by [zeDeviceGetMemoryProperties](https://spec.oneapi.io/level-zero/latest/core/api.html#ze__api_8h_1aa06c9af9880d77eaf787576df0f624d2).). So complete implementation so query for [zeDeviceGetMemoryProperties](https://spec.oneapi.io/level-zero/latest/core/api.html#ze__api_8h_1aa06c9af9880d77eaf787576df0f624d2)., determine which memory types are available, and select only those here to query for memory available (@smaslov-intel  could you confirm you get the same types from zeDeviceGetMemoryProperties and from zesMemoryGetProperties?)

However,  I dont think we have that kind of support available currently on our GPUs, with multiple memory types on the same device, so implementation on this patch is cleaner and good, but if we want to be completely strict, we would have to implement as indicated above.

Reply:
> that is actually selected by users. In the zeMemAllocDevice descriptor you have an ordinal:

> https://spec.oneapi.io/level-zero/latest/core/api.html#_CPPv4N26ze_device_mem_alloc_desc_t7ordinalE

> The ordinal is supposed to be used for this, so if you have in your system HBM and DDR5 on the same devicer, then you would use the ordinal to select one or the other


Thanks @jandres742 , this is useful. So today SYCL RT always uses ordinal 0, thus a conforming free memory query should return that for the memory type enumerated with ordinal 0 only, right? Note, that there is no way for a SYCL user to choose another memory, so that would need to be an interface extension at some point.

Reply:
@smaslov-intel Correct on both statements on ordinal and extension.

Reply:
> @smaslov-intel Correct on both statements on ordinal and extension.

Thanks @jandres742 .
@ph0b : would you rework the patch to do this:
> conforming free memory query should return that for the memory type enumerated with ordinal 0 only

Reply:
done with de96caee66dcac5e31ce57c2d2ff357ba755d00b